pkg/trace/api/otlp.go

```diff
@@ -0,0 +1,507 @@
+		Start:    int64(in.StartTimeUnixNano),
+		Duration: int64(in.EndTimeUnixNano) - int64(in.StartTimeUnixNano),
+		Service:  rattr["service.name"],
+		Resource: in.Name,
+		Meta:     rattr,
+		Metrics: map[string]float64{
+			// auto-keep all incoming traces; it was already chosen as a keeper on
+			// the client side.
+			sampler.KeySamplingPriority: float64(sampler.PriorityAutoKeep),
+		},
+	}
+	if features.Has("otlp_original_ids") {
+		// keep original IDs
+		span.Meta["otlp_ids.trace"] = hex.EncodeToString(in.TraceId)
+		span.Meta["otlp_ids.span"] = hex.EncodeToString(in.SpanId)
+		span.Meta["otlp_ids.parent"] = hex.EncodeToString(in.ParentSpanId)
+	}
+	if _, ok := span.Meta["version"]; !ok {
+		if ver := rattr["service.version"]; ver != "" {
+			span.Meta["version"] = ver
+		}
+	}
+	if len(in.Events) > 0 {
+		span.Meta["events"] = marshalEvents(in.Events)
+	}
+	for _, kv := range in.Attributes {
+		switch v := kv.Value.Value.(type) {
+		case *otlppb.AnyValue_DoubleValue:
+			span.Metrics[kv.Key] = v.DoubleValue
+		case *otlppb.AnyValue_IntValue:
+			span.Metrics[kv.Key] = float64(v.IntValue)
+		default:
+			span.Meta[kv.Key] = anyValueString(kv.Value)
+		}
+	}
+	if _, ok := span.Meta["env"]; !ok {
+		if env := span.Meta[string(semconv.DeploymentEnvironmentKey)]; env != "" {
+			span.Meta["env"] = env
+		}
+	}
+	if in.TraceState != "" {
+		span.Meta["trace_state"] = in.TraceState
+	}
+	if lib.Name != "" {
+		span.Meta["instrumentation_library.name"] = lib.Name
+	}
+	if lib.Version != "" {
+		span.Meta["instrumentation_library.version"] = lib.Version
+	}
+	if svc := span.Meta[string(semconv.PeerServiceKey)]; svc != "" {
+		span.Service = svc
+	}
+	if r := resourceFromTags(span.Meta); r != "" {
+		span.Resource = r
+	}
+	span.Type = spanKind2Type(in.Kind, span)
+	status2Error(in.Status, in.Events, span)
+	return span
+}
+
+// resourceFromTags attempts to deduce a more accurate span resource from the given list of tags meta.
+// If this is not possible, it returns an empty string.
+func resourceFromTags(meta map[string]string) string {
+	var r string
+	if m := meta["http.method"]; m != "" {
+		r = m
+		if route := meta["http.route"]; route != "" {
+			r += " " + route
+		} else if route := meta["grpc.path"]; route != "" {
+			r += " " + route
+		}
+	} else if m := meta["messaging.operation"]; m != "" {
+		r = m
+		if dest := meta["messaging.destination"]; dest != "" {
+			r += " " + dest
+		}
+	}
+	return r
+}
+
+// status2Error checks the given status and events and applies any potential error and messages
+// to the given span attributes.
+func status2Error(status *otlppb.Status, events []*otlppb.Span_Event, span *pb.Span) {
+	if status == nil || status.Code != otlppb.Status_STATUS_CODE_ERROR {
+		return
+	}
+	span.Error = 1
+	for _, e := range events {
+		if strings.ToLower(e.Name) != "exception" {
+			continue
+		}
+		for _, attr := range e.Attributes {
+			switch attr.Key {
+			case "exception.message":
```

Comment:
Here you can use `ExceptionMessageKey`, `ExceptionTypeKey` and `ExceptionStacktraceKey`